### ISSUE TITLE
Slider with custom style leaves spots outside of its bounds

### ISSUE THREAD

**Issue by llfab** [AvaloniaUI/Avalonia#14270]
## Describe the bug
Slider with custom style leaves spots outside of its bounds when being dragged. See below video. I have tested that the "spots" are outside of the sliders bounds. We have had a similar issue in the past:
https://github.com/AvaloniaUI/Avalonia/issues/9659

## To Reproduce

1. Load repro app from https://github.com/llfab/Samples/tree/main/AvaloniaAnimationTest
2. Build and run
3. Look at slider before draggin it
4. Start dragging slider and observe left side

## Expected behavior
There should not be any slider colored pixels outside of its bounds

## Video

https://github.com/AvaloniaUI/Avalonia/assets/31031996/49f5c4b3-7e3e-422f-bbaf-11d0f47c77c7

## Environment 

- OS: Windows 11
- Avalonia-Version: 11.0.6

**Comment by robloo** [comment]
Two thoughts here:

 1. We had an old issue where there were rendering artifacts. At the time it was thought to be caused by Skia and was most often visible with borders. See: https://github.com/AvaloniaUI/Avalonia/issues/10873#issuecomment-1493046032 (I can't find the original issue for it, perhaps it was a discussion)
 2. Border rendering with certain BorderThickness and CornerRadius is now known to have issues. I found this with BackgroundSizing: https://github.com/AvaloniaUI/Avalonia/pull/14048#issue-2057771245

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/f1ddc13e-1c45-427d-81be-67d2f47b137e)

Checking the Fluent Slider theme it does indeed use Border. I suspect this is related to one of the two areas above.

**Comment by llfab** [comment]
Thanks for the feedback. ...and indeed the style I created uses corner radius. So any thoughts and how to work around that?

**Comment by robloo** [comment]
On the off chance it's related to the border rendering issues can you try the BackgroundSizing PR package referenced above? That would eliminate that option fairly quickly if you still see an issue.

**Comment by robloo** [comment]
I realize this does not appear at first. It is an artifact that occurs later on. That means it likely isn't a border rendering issue due to incorrect geometries (which would appear always) and is instead likely case 1 mentioned above.

Also notice that it gets worse the more the control redraws itself -- so transparency is involved. It is likely caused by SkiaSharp anti-aliasing outside expected bounds that aren't invalidated. This keeps stacking some of the anit-aliased pixels so they appear more and more during redraws until its a solid pixel.

You can likely work around it by adjusting the anti-aliasing on the border in the slider's control template. See https://github.com/AvaloniaUI/Avalonia/pull/9584

It's either a bug in SkiaSharp that will go away with 3.0 -- or we have to specially adjust the drawing area invalidation code to account for more than just the bounds of the control. It also needs to account for Skia's anti-aliasing algorithm which currently draws outside what is expected. I suspect something is already done like this for box shadows.

This is all a best guess but I've been watching this class of issue for a while trying to understand it better.

**Comment by llfab** [comment]
Thanks for this analysis. Would you recommend to set `RenderOptions.EdgeMode="Aliased"` ?

**Comment by robloo** [comment]
> Thanks for this analysis. Would you recommend to set RenderOptions.EdgeMode="Aliased" ?

Only if you can tolerate the pixilation in the rounded corners. I'm not sure which looks worse for your scenario.

---

I also see a case of this in the current ControlCatalog. Notice when the ColorPicker flyout is open, and the color is changing, the selected color (in the DropDownButton content area) gets the artifacts. These artifacts accumulate slowly over time and only occur around the corner radius. As soon as the flyout closes the UI is invalidated properly and the artifacts go away on redraw.

https://github.com/AvaloniaUI/Avalonia/assets/17993847/fc35cf40-7278-4525-a07d-34ad83ace92c

**Comment by llfab** [comment]
I don't believe we can tolerate Aliased. So, we could either do a workaround and try to invalidate the background panel once the slider changes (how would that be done?) or we sit and wait if you think this will solve itself during the next 12 months. What do you think?

**Comment by Gillibald** [comment]
Could you try playing with `RequiresFullOpacityHandling` that creates a full layer for the sub tree

**Comment by llfab** [comment]
> Could you try playing with `RequiresFullOpacityHandling` that creates a full layer for the sub tree

For the slider or the parent panel?

**Comment by robloo** [comment]
Another idea is to use a clipping geometry. That might be able to hide the issue but I haven't tried it myself.

**Comment by llfab** [comment]
Double checked:
- `RequiresFullOpacityHandling` does not change the behavior. Neither for the surrounding panel nor for the slider
- `RenderOptions.EdgeMode="Aliased"` prevents the problem but indeed the corner radius looks rougher

**Comment by robloo** [comment]
@llfab Good to know, I think the only possible work-around with no visible downsides is a special clipping geometry.

**Comment by llfab** [comment]
> @llfab Good to know, I think the only possible work-around with no visible downsides is a special clipping geometry.

Good. Any kind of coding hints on how to do that?

**Comment by timunie** [comment]
@llfab tbh I would just add another Panel blow the Border which has 1px of Margin. 

Hope this sketch makes it a bit more clear:

![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/e441caf4-9726-4607-823d-703ad24a62d9)


```diff
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="AvaloniaAnimationTest.AnimationControl"
             Padding="4">
+ <Panel Background="Tansparent" Padding="1">
    <Border Name="animatedBorderHost" 
            Background="DarkGray" 
            CornerRadius="2"
            ClipToBounds="True"
            Height="8"
            HorizontalAlignment="Stretch">
      <Border Name="innerBorder"
              Background="Green"
              CornerRadius="4"
              Width="200"
              HorizontalAlignment="Left"/>
    </Border>
+ </Panel>
</UserControl>
```

**Comment by llfab** [comment]
That did not work. The only workaround that seems to work is:

                        <Border x:Name="_myBorder" 
                                Background="Transparent"
                                Padding="2 0 0 0">
                            <Slider x:Name="_mySlider" 
                                    Classes="largeSize"
                                    Orientation="Horizontal"
                                    HorizontalAlignment="Left"
                                    VerticalAlignment="Center"
                                    ValueChanged="OnSliderValueChanged"
                                    Minimum="-1"
                                    Maximum="1"
                                    TickPlacement="TopLeft"
                                    Ticks="0"/>
                        </Border>


        private void OnSliderValueChanged(object sender, RangeBaseValueChangedEventArgs e)
        {
            _myBorder.InvalidateVisual();
        }

**Comment by llfab** [comment]
But placing this everywhere in the UI seems cumbersome. Hence, is you feeling that this will be solved within the this year?

**Comment by timunie** [comment]
We marked the issue with priortiy, to I hope we will be able to find a solution soonish. I had a similar issue where I needed to infalte dirty rect by some px. While this could also solve this, it will add more work on the renderer where not needed. So maybe we can find a better solution here.

**Comment by llfab** [comment]
Great. From my side I would love to see this solved even if I can live with a couple more months...

**Comment by kekekeks** [comment]
It seems to be a Skia bug. `SKCanvas.DrawRoundRect` seems to be somewhat ignoring the clip bounds.

I. e.

```cs
        private SKRoundRect _rect;
        private SKPaint _paint = new();

        public ClipTest()
        {
            _rect = new SKRoundRect();
            var skArray = new SKPoint[4];
            skArray[0].X = 8;
            skArray[0].Y = 8;
            skArray[1].X = 0;
            skArray[1].Y = 0;
            skArray[2].X =0;
            skArray[2].Y = 0;
            skArray[3].X =8;
            skArray[3].Y = 8;
            _rect.SetRectRadii(new SKRect(0, 0, 40, 40), skArray);
        }
        
        protected override void SkiaRender()
        {
            _paint.IsAntialias = true;
            _paint.Color = SKColors.Aqua;
            Canvas.SetMatrix(SKMatrix.CreateTranslation(200, 200));
            for (var c = 0; c < 255; c++)
            {
                Canvas.Save();
                Canvas.ClipRect(_rect.Rect);
                Canvas.Clear(SKColors.Black);
                _paint.Color = new SKColor(255, (byte)c, 255);
                Canvas.DrawRoundRect(_rect, _paint);
                Canvas.Restore();
            }
        }
```

produces this:
![image](https://github.com/AvaloniaUI/Avalonia/assets/1067584/4b362d01-791a-4d56-97c0-6020c9c9d61f)

**Comment by llfab** [comment]
> It seems to be a Skia bug. `SKCanvas.DrawRoundRect` seems to be somewhat ignoring the clip bounds.
> 
> I. e.
> 
> ```cs
>         private SKRoundRect _rect;
>         private SKPaint _paint = new();
> 
>         public ClipTest()
>         {
>             _rect = new SKRoundRect();
>             var skArray = new SKPoint[4];
>             skArray[0].X = 8;
>             skArray[0].Y = 8;
>             skArray[1].X = 0;
>             skArray[1].Y = 0;
>             skArray[2].X =0;
>             skArray[2].Y = 0;
>             skArray[3].X =8;
>             skArray[3].Y = 8;
>             _rect.SetRectRadii(new SKRect(0, 0, 40, 40), skArray);
>         }
>         
>         protected override void SkiaRender()
>         {
>             _paint.IsAntialias = true;
>             _paint.Color = SKColors.Aqua;
>             Canvas.SetMatrix(SKMatrix.CreateTranslation(200, 200));
>             for (var c = 0; c < 255; c++)
>             {
>                 Canvas.Save();
>                 Canvas.ClipRect(_rect.Rect);
>                 Canvas.Clear(SKColors.Black);
>                 _paint.Color = new SKColor(255, (byte)c, 255);
>                 Canvas.DrawRoundRect(_rect, _paint);
>                 Canvas.Restore();
>             }
>         }
> ```
> 
> produces this: ![image](https://private-user-images.githubusercontent.com/1067584/302572032-4b362d01-791a-4d56-97c0-6020c9c9d61f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMjYzNzMsIm5iZiI6MTcwNzIyNjA3MywicGF0aCI6Ii8xMDY3NTg0LzMwMjU3MjAzMi00YjM2MmQwMS03OTFhLTRkNTYtOTdjMC02MDIwYzljOWQ2MWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDZUMTMyNzUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzAwZmE5YjFjZTZjNDFhOTdhNTYyNmE0YjUzNjdmZDRjMWZmZjM1ZTZhZjNmNGM5ZDE2Zjc3ZDIzNDdkNjQ5YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.00jLQhnl35AulHG8qpH_7USPoblyr2FzxnuHrv_StNY)

Thanks Nikita. Do you think Skia or SkiaSharp bug.

**Comment by llfab** [comment]
And: would that be fixed with SkiaSharp 3.x and the according Skia version?

**Comment by maxkatz6** [comment]
No, it wasn't fixed in SkiaSharp 3.0

**Comment by kekekeks** [comment]
It goes away if we use an intermediate texture and render it with clipping, but it would require some infrastructure changes.

**Comment by llfab** [comment]
Thanks Nikita for checking. Expensive but temporary workarounds may not be the best way. Question is whether it's Skia or SkiaSharp and whether we could make either team fix the root cause within say the next 8 months or so...

**Comment by mattleibow** [comment]
If this is a skia bug, can you confirm with skia fiddle and mayne file a bug with them? Mayne it is fixed in a later skia? I could do another native bump...

**Comment by kekekeks** [comment]
```cpp
void draw(SkCanvas* canvas) {
    SkPaint paint;
    paint.setAntiAlias(true);
    SkPath path;
    SkRRect rrect;
    SkVector corners[] = {{8,8}, {0,0}, {0,0}, {8,8}};
    rrect.setRectRadii({0, 0, 40, 40}, corners);
    canvas->translate(20,20);

    for(int c=0; c<255; c++)
    {
      canvas->save();
      canvas->clipRect(SkRect::MakeWH(40, 40), true);
      canvas->clear(SkColorSetARGB(0xff, 0xFF, 0xff, 0xff)); 
      canvas->drawRRect(rrect, paint); 
      canvas->restore();
    }
}
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/1067584/74c7958f-646e-411a-944b-f7d77f5fe00f)
![image](https://github.com/AvaloniaUI/Avalonia/assets/1067584/c7b7e7fa-04c9-4f4f-b182-474f5da0098f)

**Comment by kekekeks** [comment]
https://issues.skia.org/issues/327877721

**Comment by grokys** [comment]
Should have been fixed by #14806.
